keps/sig-cluster-lifecycle/kubeadm/3929-no-cri-socket-annotation/README.md

```diff
@@ -0,0 +1,756 @@
+"init upload-config" phase in Kubeadm. This annotation is used to specify the CRI
+socket endpoint used by the kubelet on each node for communication with the container
+runtime. Instead of relying on this annotation, the proposal suggests using a global
+kubelet configuration with a CRI socket specified, as well as providing the ability to
+override this configuration during kubeadm join using the --config flag. This would
+eliminate the need for kubeadm to store CRI socket configuration on each node, and
+instead rely on the Kubernetes configuration files for specifying this information.
+
+## Motivation
+
+Currently, kubeadm adds a CRI socket annotation to the Node object during the
+"init upload-config" phase, which specifies the endpoint for the CRI that is being
+used by the kubelet on each node. This annotation is persistent on the Node object,
+even if the kubelet is updated or the CRI is changed.
+
+After migration of container runtime endpoint flag to kubelet config, we can set
+cri socket in kubelet configuration.
+
+### Goals
+
+- Remove the use of CRI socket annotation on Node object
+- We will prioritize respecting the local setting over the global one.
+- [Not decided yet] For further node customized kubelet configuration, it can be saved locally
+  on disk with file path `/var/lib/kubelet/kubeadm-config.yaml`. (If not used, I will move it to Non-Goals)
+
+### Non-Goals
+
+- Update the CRI socket annotation on Node object to be the latest
+
+## Proposal
+
+1. init: upload a global kubelet configuration with cri socket
+2. join: can override it using --config
+3. upgrade: re-download global one, but should use local kubelet configuration firstly
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+## Design Details
+
+### init: upload a global kubelet configuration with cri socket
+
+- `kubeadm init` will not add the annotation to node.
+- `kubeadm init` will check the customized `--config` at first and if no cri socket is set, it will
+  auto-detect it and save it global configuration and local as well.
+
+### join: can override it using --config
+
+- `kubeadm join` will not add the annotation to node.
+- `kubeadm join` will download the kubelet configuration from apiserver and the customized `--config`
+  at first and auto-detect will work only if not set. Auto-detect may log a warning message if it may
+  be misconfigured and log a general debug log if there is multi CRI-sockets.
+
+### upgrade: re-download global one, but should use local kubelet configuration firstly
+
+- `kubeadm upgrade` will download the kubelet configuration from apiserver and respect local one.
+
+### Proposal 1: respect a list of configuration in local kubelet configuration, and in v1.27, CRI socket is the only one
+
+During `kubeadm ugprade`, kubeadm will read the local kubelet configuration in `/var/lib/kubelet/config.yaml`.
```

Comment:
Does the config file here support to config the `containerRuntimeEndpoint`? I tried to bootstrap a cluster with `--cri-socket` option, but this config is not persistented in the `/var/lib/kubelet/config.yaml`.  If the endpoint is already supported to config in kubelet config file, then this is the logic we should update.

Reply:
> Does the config file here support to config the `containerRuntimeEndpoint`? I tried to bootstrap a cluster with `--cri-socket` option, but this config is not persistented in the `/var/lib/kubelet/config.yaml`. If the endpoint is already supported to config in kubelet config file, then this is the logic we should update.

containerRuntimeEndpoint is added in v1.27. In the past and also v1.27, we are using the `--container-runtime-endpoint` in `/var/lib/kubelet/kubeadm-flags.env`. 
- Before v1.27, it cannot be added in the kubeletConfiguration.
- And also, we saved the cri-socket in the annotation, and that is why we don't have to read and parse `/var/lib/kubelet/kubeadm-flags.env` here. 

If we start the process that not using cri-socket annotation in v1.28, we set it in the `/var/lib/kubelet/config.yaml`.
Then the flag in  `/var/lib/kubelet/kubeadm-flags.env` can be cleaned. (This should be mentioned in this KEP.  I will update it.)